server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -154,7 +154,11 @@ public void allocate(RoutingAllocation allocation) {
         final BalancingWeights balancingWeights = balancingWeightsFactory.create();
         final Balancer balancer = new Balancer(writeLoadForecaster, allocation, balancerSettings.getThreshold(), balancingWeights);
         balancer.allocateUnassigned();
-        balancer.moveShards();
+
+        // TODO: if (balancer.moveNonPreferred()) return;
+        if (balancer.moveShards()) {
+            return;
```

Comment:
If we return early here we'll skip `collectAndRecordNodeWeightStats` below. Perhaps we need to revisit that behaviour now that we'll only be doing a single move at a time.

Reply:
In dbd0fede791b9a31aa750108d32f7d9b99775d18, I simiply moved `collectAndRecordNodeWeightStats` into `finally` block so that it always runs. We can revisit whether it can be skipped in some situations which can be handled separately from this PR imo.